### PR TITLE
Issue 579 Don't start heartbeat in createSocket

### PR DESCRIPTION
**Pull request by davedoesdev**
#579 

Test uses `process._getActiveHandles()` to check `setInterval()` wasn't called.
Alternative would be to mock `setInterval`.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1093,6 +1093,10 @@ Primus.prototype.reserved.events = {
  */
 Primus.createSocket = function createSocket(options) {
   options = options || {};
+  // Make sure the temporary Primus we create below doesn't start a heartbeat
+  options.pingInterval = 'pingInterval' in options
```
Comment (lpinca) [inline index.js]: Should we set it to `false` unconditionally? Can't think of a case where starting the heartbeat timer makes sense here.

File: test/transformer.base.js [hunk 2]
```diff
@@ -898,6 +898,22 @@ module.exports = function base(transformer, transformer_name) {
...
       });
+
+      it('should not start heartbeat', function (done) {
+        var num_handles_before = process._getActiveHandles().length
+          , PSocket = Primus.createSocket({
+              transformer: transformer,
+              pathname: server.pathname
+            })
+          , num_handles_after = process._getActiveHandles().length
+          , socket = new PSocket(server.addr);
```
Comment (lpinca) [inline test/transformer.base.js]: If I'm not wrong this is not needed for this test so I would remove it.

**Comment by davedoesdev** [comment]
I've reworked the test to check `setInterval` isn't called rather than rely on the private function `process._getActiveHandles`.

I've also set `pingInterval` to `false` always. However, I'm modifying the passed-in `options` parameter which I'm a little uncomfortable with. I notice other places in `index.js` do this too. Wouldn't it be better to extend `options` instead of modifying what the caller passes in?

**Comment by lpinca** [comment]
Yes I agree, we should not mutate the original options object. There have been no complaints or bug reports though.

File: index.js [hunk 3]
```diff
@@ -1092,7 +1092,9 @@ Primus.prototype.reserved.events = {
  * @api public
  */
 Primus.createSocket = function createSocket(options) {
-  options = options || {};
+  options = Primus.prototype.merge.call(Primus, {}, options || {});
```
Comment (lpinca) [inline index.js]: We can also safely use `Object.assign()`.

Comment (davedoesdev) [inline index.js]: Yes, I normally do that but didn't find it's use anywhere else in that file.

Comment (lpinca) [inline index.js]: It's used in the server code of the transformers (e.g. https://github.com/primus/primus/blob/509cf0294728a5b2000b825d40c56560b0f4e5c6/transformers/websockets/server.js#L15) so feel free to use it.

**Comment by tejohnso** [comment]
This change breaks the heartbeat protocol. The client no longer disconnects/reconnects if a ping is not received within the timeout. 

> You might need to connect from a different node process where you don't have access to your primus instance and the compatible Socket instance. For these cases there a special createSocket method ...

There's no indication that processes created with the `createSocket` method should automatically exit when the connection is broken. The documentation suggests that the `createSocket` client behaves the same as any other. 

Maybe that's not the intention, but if that's the case, I think this should at least have been part of a major version change, and the documentation should be updated to clearly indicate that clients created through `createSocket` don't behave the same as clients created by the "from the browser" method.

Is the intention not to use `createSocket` in node processes?

**Pull request merged**